### DISCUSSION TITLE
graphql POST request results in 404

### DISCUSSION THREAD

**Discussion by DennisKraaijeveld** [Shopify/shopify-app-template-node#1048] [Q&A]
Hi people, 

I have built a basic app to setup pre-orders and I am using the Apollo client to make the POST calls for actually submitting those rules. But every time I want to send the data, I will get a 404 error in my console. The url is a NGROK tunnel i.e; https://3z9a-sndsa-xxxx-xxxx/graphql

Am I doing something wrong?

**Comment by sgarner** [comment]
Same problem here.

The graphql proxy functionality was removed from the template in https://github.com/Shopify/shopify-app-template-node/pull/935/commits/743133034ff241a69f4f52ddd6abfcb354e47e4a, but there is no indication of what should replace it.

As a result, tutorials in the Shopify documentation [like this one](https://shopify.dev/apps/purchase-options/deferred/create-app/getting-started) no longer work.

**Comment by fonya2014** [comment]
Hi! I am facing the same issue and I am recreating all my queries work through fetch and Shopify.Clients.Graphql. Here is an example:

https://github.com/Shopify/shopify-frontend-template-react/blob/9d2caa0c416037cfbf38824169f0f06b350372c7/components/ProductsCard.jsx

https://github.com/Shopify/shopify-app-template-node/blob/e678d844825d9893d77a6c7f68fcee26795cde74/web/helpers/product-creator.js

**Reply by sgarner** [reply]
Yes, the only solution I've found is to create endpoints that perform the Admin GraphQL queries on the server-side and then access those endpoints from the frontend - instead of using the Shopify Admin API directly. But this seems like a step backwards compared with the proxy functionality.

**Comment by mdj-uk** [comment]
Has there ever been an explanation as to why the graphql proxy was removed? The only hint I can see is on PR https://github.com/Shopify/shopify-app-template-node/pull/908 which says it's company policy to 'prefer offline to online session tokens' and links to some non-public issue tracker. 

So for the rest of us we're none the wiser. 

Perhaps there's a good reason, but it does seem that the consequence is to make our apps more complicated than they need to be.

**Comment by mkevinosullivan** [comment]
The [GraphQL functionality](https://github.com/Shopify/shopify-api-js/blob/main/docs/reference/clients/graphqlProxy.md) is still in the library.

The proxy functionality was removed from the template at the same time as changing the default sessions to offline because the GraphQL proxy functionality can _only_ work with online sessions.

**Comment by danyn** [comment]
What mkevinsullivan says makes sense.
Still is there not a code example for how to use the proxy for online sessions?  
It reduces the amount of code needed, but introduces the problem of needing the raw data from the request.
I started to approach adding this back.  It seemed to be just adding the middleware in express to get the raw body when using the correct header. 

The issue seemed to lack any good docs from shopify as it is a general back end problem.  So I avoided it due to lack of certainty.
But out of respect to the fact that the app templates are changing in breaking ways so often it would be nice to include a code example for the proxy as it used to be the default of handling gql when the package still used apollo.
